File: apps/api/app/simulation.py
"""
Monte Carlo simulation engine – ported from simulation.js.
Returns the full simulation result in one call.
"""
import math
import random
from typing import Optional

from sqlmodel import Session, select
from app.models import Asset, Event, Tip, Question, Decision


def gaussian_random(mean: float, std_dev: float) -> float:
    u = 0.0
    v = 0.0
    while u == 0:
        u = random.random()
    while v == 0:
        v = random.random()
    z = math.sqrt(-2.0 * math.log(u)) * math.cos(2.0 * math.pi * v)
    return mean + z * std_dev


def run_simulation(
    session: Session,
    years: int,
    total_budget: float,
    allocation: dict,
) -> dict:
    """
    Run the full market simulation.
    allocation: { asset_id: percentage, ... } – must sum to ~100.
    """
    # Load data from DB
    assets_list = session.exec(select(Asset)).all()
    events_list = session.exec(select(Event)).all()
    tips_list = session.exec(select(Tip)).all()
    questions_list = session.exec(select(Question)).all()
    decisions_list = session.exec(select(Decision)).all()

    asset_map = {a.id: a for a in assets_list}

    # Build portfolio
    portfolio = {}
    total_pct = 0.0
    for asset_id, pct in allocation.items():
        pct = float(pct)
        if pct > 0 and asset_id in asset_map:
            a = asset_map[asset_id]
            amount = total_budget * (pct / 100.0)
            portfolio[asset_id] = {
                "asset": a,
                "percentage": pct,
                "amount": amount,
                "history": [amount],
            }
            total_pct += pct

    if total_pct < 99 or total_pct > 101:
        return {"error": f"Allocation must sum to 100%. Current: {total_pct}%"}

    # Pick market events ~ every 5 years
    shuffled_events = list(events_list)
    random.shuffle(shuffled_events)
    event_years = {}
    event_idx = 0
    for y in range(5, years + 1, 5):
        if event_idx < len(shuffled_events):
            event_years[y] = shuffled_events[event_idx]
            event_idx += 1

    # Pre-generate yearly interactive events
    shuffled_tips = list(tips_list)
    random.shuffle(shuffled_tips)
    shuffled_questions = list(questions_list)
    random.shuffle(shuffled_questions)
    shuffled_decisions = list(decisions_list)
    random.shuffle(shuffled_decisions)
    tip_idx = q_idx = d_idx = 0

    yearly_interactions: dict = {}
    for y in range(1, years + 1):
        roll = random.random()
        if roll < 0.33 and tip_idx < len(shuffled_tips):
            t = shuffled_tips[tip_idx]
            tip_idx += 1
            yearly_interactions[y] = {
                "type": "tip",
                "data": {"id": t.id, "icon": t.icon, "title": t.title, "text": t.text},
            }
        elif roll < 0.66 and q_idx < len(shuffled_questions):
            q = shuffled_questions[q_idx]
            q_idx += 1
            yearly_interactions[y] = {
                "type": "quiz",
                "data": {
                    "id": q.id,
                    "question": q.question,
                    "options": q.options,
                    "correct": q.correct,
                    "explanation": q.explanation,
                },
            }
        elif d_idx < len(shuffled_decisions):
            d = shuffled_decisions[d_idx]
            d_idx += 1
            yearly_interactions[y] = {
                "type": "decision",
                "data": {
                    "id": d.id,
                    "title": d.title,
                    "icon": d.icon,
                    "description": d.description,
                    "options": d.options,
                    "lesson": d.lesson,
                },
            }
        elif tip_idx < len(shuffled_tips):
            t = shuffled_tips[tip_idx]
            tip_idx += 1
            yearly_interactions[y] = {
                "type": "tip",
                "data": {"id": t.id, "icon": t.icon, "title": t.title, "text": t.text},
            }
        elif q_idx < len(shuffled_questions):
            q = shuffled_questions[q_idx]
            q_idx += 1
            yearly_interactions[y] = {
                "type": "quiz",
                "data": {
                    "id": q.id,
                    "question": q.question,
                    "options": q.options,
                    "correct": q.correct,
                    "explanation": q.explanation,
                },
            }
        else:
            # Fallback: random tip
            t = random.choice(tips_list)
            yearly_interactions[y] = {
                "type": "tip",
                "data": {"id": t.id, "icon": t.icon, "title": t.title, "text": t.text},
            }

    # ── Simulate year by year ─────────────────────────────────────────
    yearly_data = []
    total_portfolio_history = [total_budget]
    saver_amount = total_budget
    saver_history = [total_budget]
    saver_rate = 0.005

    trader_amount = total_budget
    trader_history = [total_budget]
    trader_fee_rate = 0.02
    inflation_rate = 0.02
    panic_sells = 0
    triggered_events = []

    for year in range(1, years + 1):
        year_event = event_years.get(year)
        total_value = 0.0
        year_returns = {}

        for asset_id, pos in portfolio.items():
            asset = pos["asset"]
            year_return = gaussian_random(asset.avg_return, asset.volatility)

            if year_event:
                impact = (year_event.impacts or {}).get(asset.category, 0)
                year_return += impact

            year_return = max(year_return, -0.60)
            year_return = min(year_return, 1.50)

            pos["amount"] = pos["amount"] * (1 + year_return)
            pos["amount"] = max(pos["amount"], 0)
            pos["history"].append(pos["amount"])
            year_returns[asset_id] = year_return
            total_value += pos["amount"]

        # Saver benchmark
        saver_amount *= 1 + saver_rate
        saver_history.append(saver_amount)

        # Trader benchmark
        trader_return = gaussian_random(0.06, 0.28)
        if year_event:
            trader_return -= 0.15
        trader_return = max(trader_return, -0.60)
        trader_return = min(trader_return, 1.50)
        trader_amount = trader_amount * (1 + trader_return) * (1 - trader_fee_rate)
        trader_amount = max(trader_amount, 0)
        trader_history.append(trader_amount)

        previous_total = total_portfolio_history[-1]
        year_change = (total_value - previous_total) / previous_total if previous_total else 0
        if year_change < -0.15:
            panic_sells += 1

        total_portfolio_history.append(total_value)

        if year_event:
            triggered_events.append({
                "year": year,
                "id": year_event.id,
                "title": year_event.title,
                "description": year_event.description,
                "lesson": year_event.lesson,
                "icon": year_event.icon,
                "impacts": year_event.impacts,
            })

        interaction = yearly_interactions.get(year)

        event_data = None
        if year_event:
            event_data = {
                "title": year_event.title,
                "icon": year_event.icon,
                "description": year_event.description,
                "lesson": year_event.lesson,
            }

        yearly_data.append({
            "year": year,
            "totalValue": round(total_value),
            "yearReturn": year_change,
            "event": event_data,
            "interaction": interaction,
            "assetReturns": year_returns,
        })

    # ── Final stats ───────────────────────────────────────────────────
    final_value = total_portfolio_history[-1]
    total_return = final_value - total_budget
    total_return_pct = (total_return / total_budget) * 100

    unique_categories = set(pos["asset"].category for pos in portfolio.values())
    diversification_score = min(
        100,
        round((len(unique_categories) / 5) * 60 + (len(portfolio) / 6) * 40),
    )

    cagr = (final_value / total_budget) ** (1 / years) - 1 if years > 0 else 0
    annualized_return_pct = round(cagr * 1000) / 10

    yearly_returns_list = [y["yearReturn"] for y in yearly_data]
    avg_ret = sum(yearly_returns_list) / len(yearly_returns_list) if yearly_returns_list else 0
    variance = sum((r - avg_ret) ** 2 for r in yearly_returns_list) / len(yearly_returns_list) if yearly_returns_list else 0
    vol = math.sqrt(variance)
    volatility_pct = round(vol * 1000) / 10

    risk_free_rate = 0.005
    excess_return = cagr - risk_free_rate
    sharpe_ratio = round((excess_return / vol) * 100) / 100 if vol > 0 else 0

    if sharpe_ratio >= 0.8:
        risk_return_grade = "A"
    elif sharpe_ratio >= 0.5:
        risk_return_grade = "B"
    elif sharpe_ratio >= 0.2:
        risk_return_grade = "C"
    elif sharpe_ratio >= 0:
        risk_return_grade = "D"
    else:
        risk_return_grade = "F"

    weighted_risk = sum(
        pos["asset"].volatility * pos["percentage"] / 100 for pos in portfolio.values()
    )
    if weighted_risk > 0.25:
        risk_level = "Aggressive"
    elif weighted_risk > 0.15:
        risk_level = "Moderate"
    else:
        risk_level = "Conservative"

    # Investor score
    investor_score = 0
    if total_return > 0:
        investor_score += 15
    if total_return_pct > 20:
        investor_score += 5
    investor_score += round(diversification_score * 0.30)
    investor_score += max(0, min(30, round(sharpe_ratio * 30)))
    investor_score += max(0, 20 - panic_sells * 6)
    investor_score = min(100, max(0, investor_score))

    cost_of_fear = round(abs(total_return) * 0.15 * panic_sells)

    lessons = []
    if total_return > 0:
        lessons.append({"icon": "📊", "text": "Markets go up and down — that's normal, not failure."})
    lessons.append({"icon": "⏰", "text": "Time in the market beats timing the market."})
    if len(unique_categories) >= 3:
        lessons.append({"icon": "🎯", "text": "Diversification reduces your risk without killing returns."})
    if panic_sells > 0:
        lessons.append({"icon": "💪", "text": "Staying invested during crashes is the hardest — and smartest — move."})
    if total_return > saver_amount - total_budget:
        lessons.append({"icon": "🚀", "text": "Investing beats saving — your money works harder for you."})
    if sharpe_ratio < 0.3 and weighted_risk > 0.2:
        lessons.append({"icon": "⚠️", "text": "Taking more risk didn't pay off here. A balanced portfolio often performs better risk-adjusted."})

    return {
        "summary": {
            "investorScore": investor_score,
            "finalValue": round(final_value),
            "totalReturn": round(total_return),
            "totalReturnPct": round(total_return_pct * 10) / 10,
            "saverFinalValue": round(saver_amount),
            "saverInflationLoss": round(total_budget * (1 - 1 / ((1 + inflation_rate) ** years))),
            "costOfFear": cost_of_fear,
            "panicSells": panic_sells,
            "diversificationScore": diversification_score,
            "lessons": lessons,
            "annualizedReturnPct": annualized_return_pct,
            "volatilityPct": volatility_pct,
            "sharpeRatio": sharpe_ratio,
            "riskReturnGrade": risk_return_grade,
            "riskLevel": risk_level,
        },
        "yearlyData": yearly_data,
        "portfolioHistory": [round(v) for v in total_portfolio_history],
        "saverHistory": [round(v) for v in saver_history],
        "traderHistory": [round(v) for v in trader_history],
        "assetHistories": {
            aid: {
                "name": pos["asset"].name,
                "icon": pos["asset"].icon,
                "category": pos["asset"].category,
                "history": [round(v) for v in pos["history"]],
                "percentage": pos["percentage"],
            }
            for aid, pos in portfolio.items()
        },
        "triggeredEvents": triggered_events,
        "years": years,
        "totalBudget": total_budget,
    }

File: apps/ui/server/simulation.js
// Monte Carlo-style market simulation engine with yearly interactive events
const assets = require('./data/assets');
const events = require('./data/events');
const tips = require('./data/tips');
const questions = require('./data/questions');
const decisions = require('./data/decisions');

/**
 * Run a market simulation with yearly interactive events.
 * Each year, the player encounters one of: Tip, Quiz Question, or Crisis Decision.
 */
function runSimulation({ years, totalBudget, allocation }) {
  const assetMap = {};
  assets.forEach(a => { assetMap[a.id] = a; });

  // Build portfolio from allocation
  const portfolio = {};
  let totalPct = 0;
  for (const [assetId, pct] of Object.entries(allocation)) {
    if (pct > 0 && assetMap[assetId]) {
      portfolio[assetId] = {
        asset: assetMap[assetId],
        percentage: pct,
        amount: totalBudget * (pct / 100),
        history: [totalBudget * (pct / 100)]
      };
      totalPct += pct;
    }
  }

  if (totalPct < 99 || totalPct > 101) {
    return { error: `Allocation must sum to 100%. Current: ${totalPct}%` };
  }

  // Pick market events (crises) that will occur - roughly every 5 years
  const shuffledEvents = [...events].sort(() => Math.random() - 0.5);
  const eventYears = {};
  let eventIdx = 0;
  for (let y = 5; y <= years; y += 5) {
    if (eventIdx < shuffledEvents.length) {
      eventYears[y] = shuffledEvents[eventIdx];
      eventIdx++;
    }
  }

  // Pre-generate yearly interactive events (tip, quiz, or decision)
  const shuffledTips = [...tips].sort(() => Math.random() - 0.5);
  const shuffledQuestions = [...questions].sort(() => Math.random() - 0.5);
  const shuffledDecisions = [...decisions].sort(() => Math.random() - 0.5);
  let tipIdx = 0, qIdx = 0, dIdx = 0;

  const yearlyInteractions = {};
  for (let y = 1; y <= years; y++) {
    // Distribute evenly: tip, quiz, decision in rotation with some randomness
    const roll = Math.random();
    if (roll < 0.33 && tipIdx < shuffledTips.length) {
      yearlyInteractions[y] = { type: 'tip', data: shuffledTips[tipIdx++] };
    } else if (roll < 0.66 && qIdx < shuffledQuestions.length) {
      yearlyInteractions[y] = { type: 'quiz', data: shuffledQuestions[qIdx++] };
    } else if (dIdx < shuffledDecisions.length) {
      yearlyInteractions[y] = { type: 'decision', data: shuffledDecisions[dIdx++] };
    } else if (tipIdx < shuffledTips.length) {
      yearlyInteractions[y] = { type: 'tip', data: shuffledTips[tipIdx++] };
    } else if (qIdx < shuffledQuestions.length) {
      yearlyInteractions[y] = { type: 'quiz', data: shuffledQuestions[qIdx++] };
    } else {
      // Fallback: recycle a random tip
      yearlyInteractions[y] = { type: 'tip', data: tips[Math.floor(Math.random() * tips.length)] };
    }
  }

  // Simulate year by year
  const yearlyData = [];
  let totalPortfolioHistory = [totalBudget];
  let saverAmount = totalBudget;
  const saverHistory = [totalBudget];
  const saverRate = 0.005;
  // Trader profile: aggressive short-term trader (higher volatility, fees, panic sells)
  let traderAmount = totalBudget;
  const traderHistory = [totalBudget];
  const traderFeeRate = 0.02; // 2% annual transaction costs from frequent trading
  const inflationRate = 0.02;
  let panicSells = 0;
  const triggeredEvents = [];

  for (let year = 1; year <= years; year++) {
    let yearEvent = eventYears[year] || null;
    let totalValue = 0;
    const yearReturns = {};

    for (const [assetId, pos] of Object.entries(portfolio)) {
      const asset = pos.asset;
      let yearReturn = gaussianRandom(asset.avgReturn, asset.volatility);
      
      if (yearEvent) {
        const impact = yearEvent.impacts[asset.category] || 0;
        yearReturn += impact;
      }

      yearReturn = Math.max(yearReturn, -0.60);
      yearReturn = Math.min(yearReturn, 1.50);

      pos.amount = pos.amount * (1 + yearReturn);
      pos.amount = Math.max(pos.amount, 0);
      pos.history.push(pos.amount);
      yearReturns[assetId] = yearReturn;
      totalValue += pos.amount;
    }

    saverAmount = saverAmount * (1 + saverRate);
    saverHistory.push(saverAmount);

    // Trader simulation: amplified volatility, fees, and panic selling tendency
    let traderReturn = gaussianRandom(0.06, 0.28); // higher vol, slightly lower mean
    if (yearEvent) {
      // Trader panics during events -> sells at bottom (-15% extra loss)
      traderReturn -= 0.15;
    }
    traderReturn = Math.max(traderReturn, -0.60);
    traderReturn = Math.min(traderReturn, 1.50);
    traderAmount = traderAmount * (1 + traderReturn) * (1 - traderFeeRate);
    traderAmount = Math.max(traderAmount, 0);
    traderHistory.push(traderAmount);

    const previousTotal = totalPortfolioHistory[totalPortfolioHistory.length - 1];
    const yearChange = (totalValue - previousTotal) / previousTotal;
    if (yearChange < -0.15) {
      panicSells++;
    }

    totalPortfolioHistory.push(totalValue);

    if (yearEvent) {
      triggeredEvents.push({ year, ...yearEvent });
    }

    // Get the interactive event for this year
    const interaction = yearlyInteractions[year] || null;

    yearlyData.push({
      year,
      totalValue: Math.round(totalValue),
      yearReturn: yearChange,
      // Market crisis event (for portfolio impact display)
      event: yearEvent ? { title: yearEvent.title, icon: yearEvent.icon, description: yearEvent.description, lesson: yearEvent.lesson } : null,
      // Interactive event (tip, quiz, or decision) — shown every year
      interaction: interaction,
      assetReturns: yearReturns
    });
  }

  // Calculate final results
  const finalValue = totalPortfolioHistory[totalPortfolioHistory.length - 1];
  const totalReturn = finalValue - totalBudget;
  const totalReturnPct = (totalReturn / totalBudget) * 100;
  
  const uniqueCategories = new Set(Object.values(portfolio).map(p => p.asset.category));
  const diversificationScore = Math.min(100, Math.round((uniqueCategories.size / 5) * 60 + (Object.keys(portfolio).length / 6) * 40));

  // --- RISK-RETURN RATIO ---
  // Calculate annualized return
  const cagr = Math.pow(finalValue / totalBudget, 1 / years) - 1;
  const annualizedReturnPct = Math.round(cagr * 1000) / 10;

  // Calculate portfolio volatility (std dev of yearly returns)
  const yearlyReturns = yearlyData.map(y => y.yearReturn);
  const avgReturn = yearlyReturns.reduce((a, b) => a + b, 0) / yearlyReturns.length;
  const variance = yearlyReturns.reduce((sum, r) => sum + Math.pow(r - avgReturn, 2), 0) / yearlyReturns.length;
  const volatility = Math.sqrt(variance);
  const volatilityPct = Math.round(volatility * 1000) / 10;

  // Risk-adjusted return (Sharpe-like: excess return over risk-free / volatility)
  const riskFreeRate = 0.005; // 0.5% savings rate
  const excessReturn = cagr - riskFreeRate;
  const sharpeRatio = volatility > 0 ? Math.round((excessReturn / volatility) * 100) / 100 : 0;

  // Risk-return grade (penalizes high risk for marginal returns)
  let riskReturnGrade;
  if (sharpeRatio >= 0.8) riskReturnGrade = 'A';
  else if (sharpeRatio >= 0.5) riskReturnGrade = 'B';
  else if (sharpeRatio >= 0.2) riskReturnGrade = 'C';
  else if (sharpeRatio >= 0) riskReturnGrade = 'D';
  else riskReturnGrade = 'F';

  // Weighted portfolio risk level (high volatility assets weighted by allocation %)
  const weightedRisk = Object.entries(portfolio).reduce((sum, [id, pos]) => {
    return sum + (pos.asset.volatility * pos.percentage / 100);
  }, 0);
  const riskLevel = weightedRisk > 0.25 ? 'Aggressive' : weightedRisk > 0.15 ? 'Moderate' : 'Conservative';

  // --- INVESTOR SCORE (risk-adjusted) ---
  let investorScore = 0;
  // Points for positive return (max 20)
  if (totalReturn > 0) investorScore += 15;
  if (totalReturnPct > 20) investorScore += 5;
  // Points for diversification (max 30)
  investorScore += Math.round(diversificationScore * 0.30);
  // Points for risk-adjusted return (max 30) — THIS is key: high risk with low return = bad
  investorScore += Math.max(0, Math.min(30, Math.round(sharpeRatio * 30)));
  // Points for not experiencing extreme drawdowns (max 20)
  investorScore += Math.max(0, 20 - panicSells * 6);

  investorScore = Math.min(100, Math.max(0, investorScore));

  const costOfFear = Math.round(Math.abs(totalReturn) * 0.15 * panicSells);

  const lessons = [];
  if (totalReturn > 0) {
    lessons.push({ icon: "📊", text: "Markets go up and down — that's normal, not failure." });
  }
  lessons.push({ icon: "⏰", text: "Time in the market beats timing the market." });
  if (uniqueCategories.size >= 3) {
    lessons.push({ icon: "🎯", text: "Diversification reduces your risk without killing returns." });
  }
  if (panicSells > 0) {
    lessons.push({ icon: "💪", text: "Staying invested during crashes is the hardest — and smartest — move." });
  }
  if (totalReturn > saverAmount - totalBudget) {
    lessons.push({ icon: "🚀", text: "Investing beats saving — your money works harder for you." });
  }
  if (sharpeRatio < 0.3 && weightedRisk > 0.2) {
    lessons.push({ icon: "⚠️", text: "Taking more risk didn't pay off here. A balanced portfolio often performs better risk-adjusted." });
  }

  return {
    summary: {
      investorScore,
      finalValue: Math.round(finalValue),
      totalReturn: Math.round(totalReturn),
      totalReturnPct: Math.round(totalReturnPct * 10) / 10,
      saverFinalValue: Math.round(saverAmount),
      saverInflationLoss: Math.round(totalBudget * (1 - 1 / Math.pow(1 + inflationRate, years))),
      costOfFear,
      panicSells,
      diversificationScore,
      lessons,
      // Risk-return metrics
      annualizedReturnPct,
      volatilityPct,
      sharpeRatio,
      riskReturnGrade,
      riskLevel
    },
    yearlyData,
    portfolioHistory: totalPortfolioHistory.map(Math.round),
    saverHistory: saverHistory.map(Math.round),
    traderHistory: traderHistory.map(Math.round),
    // Per-asset histories for individual asset chart
    assetHistories: Object.fromEntries(
      Object.entries(portfolio).map(([id, pos]) => [id, {
        name: pos.asset.name,
        icon: pos.asset.icon,
        category: pos.asset.category,
        history: pos.history.map(Math.round),
        percentage: pos.percentage,
      }])
    ),
    triggeredEvents,
    years,
    totalBudget
  };
}

function gaussianRandom(mean, stdDev) {
  let u = 0, v = 0;
  while (u === 0) u = Math.random();
  while (v === 0) v = Math.random();
  const z = Math.sqrt(-2.0 * Math.log(u)) * Math.cos(2.0 * Math.PI * v);
  return mean + z * stdDev;
}

module.exports = { runSimulation };
